addons/product/models/res_company.py

```diff
@@ -1,67 +1,47 @@
-                ('company_id', '=', False)
-            ], limit=1)
-            if not pricelist:
-                params = {'currency': new_company.currency_id.name}
-                pricelist = ProductPricelist.create({
-                    'name': _("Default %(currency)s pricelist") %  params,
-                    'currency_id': new_company.currency_id.id,
-                })
-            self.env['ir.property']._set_default(
-                'property_product_pricelist',
-                'res.partner',
-                pricelist,
-                new_company,
-            )
+        companies = super().create(vals_list)
+        self._manage_companies_pricelists()
         return companies
 
-    def write(self, values):
-        # When we modify the currency of the company, we reflect the change on the list0 pricelist, if
-        # that pricelist is not used by another company. Otherwise, we create a new pricelist for the
-        # given currency.
-        ProductPricelist = self.env['product.pricelist']
-        currency_id = values.get('currency_id')
-        main_pricelist = self.env.ref('product.list0', False)
-        if currency_id and main_pricelist:
-            nb_companies = self.search_count([])
-            for company in self:
-                existing_pricelist = ProductPricelist.search(
-                    [('company_id', 'in', (False, company.id)),
-                     ('currency_id', 'in', (currency_id, company.currency_id.id))])
-                if existing_pricelist and any(currency_id == x.currency_id.id for x in existing_pricelist):
-                    continue
-                if currency_id == company.currency_id.id:
-                    continue
-                currency_match = main_pricelist.currency_id == company.currency_id
-                company_match = (main_pricelist.company_id == company or
-                                 (main_pricelist.company_id.id is False and nb_companies == 1))
-                if currency_match and company_match:
-                    main_pricelist.write({'currency_id': currency_id})
-                else:
-                    params = {'currency': self.env['res.currency'].browse(currency_id).name}
-                    pricelist = ProductPricelist.create({
-                        'name': _("Default %(currency)s pricelist") %  params,
-                        'currency_id': currency_id,
-                    })
-                    self.env['ir.property']._set_default(
-                        'property_product_pricelist',
-                        'res.partner',
-                        pricelist,
-                        company,
-                    )
-        return super(ResCompany, self).write(values)
+    def _manage_companies_pricelists(self):
+        """ Manage the default pricelists for all companies. """
+        if self.user_has_groups('product.group_product_pricelist'):
+            # Activate existing default pricelists
```

Comment:
if self is non empty, only create pricelists for the companies in self.

Reply:
For both points here, I just realized something. When creating a new company, you can use an inactive currency, thus automatically activating it as well as the pricelist setting before creating the company. So I passed self to reflect the change of the setting on all the companies.

This should be done at the currency activation and not here. Right...